### DISCUSSION TITLE
Intermittent work / fail behavior of OCP CAD Viewer

### DISCUSSION THREAD

**Discussion by WayneSherman** [bernhard-42/vscode-ocp-cad-viewer#104] [Q&A]
Using
- a recently updated Manjaro (arch linux based)
- tested with both python 3.12.5 and python 3.10.14
- vscode 1.93.1
- Installed extensions
  code --install-extension ms-python.python
  code --install-extension bernhard-42.ocp-cad-viewer  # ver 2.4.1
- Used the "Quick build123d" install button

After following the installation instructions, if I directly load the auto created file "ocp_vscode_demo.py" then everything works.

But if I open "folder" which opens a "[workspace](https://code.visualstudio.com/docs/editor/workspaces#_how-do-i-open-a-vs-code-workspace)" (the folder only contains the one demo file), the viewer pane stays empty and I get the following error when trying to run ocp_vscode_demo.py:

```
build123d_vscode[test@manjaro-22-vm demo]$ /home/test/build123d_vscode/bin/python /home/test/build123d_vscode/projects/demo/ocp_vscode_demo.py
Using port 3939 taken from config file
Jupyter console not installed
Traceback (most recent call last):
  File "/home/test/build123d_vscode/lib/python3.10/site-packages/ocp_vscode/config.py", line 411, in status
    return dict(sorted(response.get("text", {}).items()))
AttributeError: 'str' object has no attribute 'get'
```
The above exception leads to other exceptions and finally "RuntimeError: Cannot access viewer config. Is the viewer running?"

Thanks for the help.

**Comment by bernhard-42** [comment]
I'm on vacation and away from any keyboard, so I cannot reproduce it.
Is the viewer running in VS Code and listening to port 3939. This can be seen when you click on the OCP CAD Viewer button in the navigation bar and there in the top viewer manager

**Reply by WayneSherman** [reply]
Thanks for your reply.  This can wait until you are back at your computer.  To answer your question:

The CAD viewer appears to be running:  Viewer Manager | ocp_vscode RUNNING, port 3939
(the OCP CAD Viewer tab is present, but it is an empty gray box with no controls or drawing in it)

Terminal Output:
```
build123d_vscode[test@manjaro-22-vm demo]$ /home/test/build123d_vscode/bin/python /home/test/.vscode/extensions/
![vscode-ocp-cad-viewer-viewer and libraries managers](https://github.com/user-attachments/assets/9f9769d7-af9a-4fcd-aaae-da7085f32eee)
![vscode-ocp-cad-viewer-open demo folder](https://github.com/user-attachments/assets/f1e0ae21-b3aa-4cfc-aee5-21c35f317a74)
bernhard-42.ocp-cad-viewer-2.4.1/ocp_vscode/backend.py --port 3939
Viewer backend started
Logo model loaded
```

OCP CAD Viewer LOG:
```
[11:07:49 AM.846} INFO  ] Successfully registered CadqueryViewer Status Manager
[11:07:49 AM.871} INFO  ] Successfully registered CadqueryViewer Library Manager
[11:07:49 AM.879} DEBUG ] PythonPath: '/home/test/build123d_vscode/bin/python', environment: VirtualEnvironment, build123d_vscode
[11:07:51 AM.850} DEBUG ] Reviving webview panel
[11:07:51 AM.852} INFO  ] ocp_vscode library version 2.4.1 matches extension version 2.4.1
[11:07:51 AM.859} DEBUG ] PythonPath: '/home/test/build123d_vscode/bin/python', environment: VirtualEnvironment, build123d_vscode
[11:07:51 AM.862} INFO  ] Server started on port 3939
[11:07:51 AM.862} INFO  ] Starting websocket server ...
[11:07:51 AM.862} DEBUG ] Creating new webview panel
[11:07:51 AM.865} DEBUG ] Updateing webview
[11:07:51 AM.866} DEBUG ] OCP Cad Viewer port: 3939, folder: /home/test/build123d_vscode/projects/demo, /home/test/build123d_vscode/projects/demo
[11:07:51 AM.868} DEBUG ] Command OCP CAD Viewer registered
[11:07:52 AM.034} DEBUG ] Lock file /home/test/.ocpvscode.lock acquired
[11:07:52 AM.160} DEBUG ] Lock file /home/test/.ocpvscode.lock removed
[11:07:52 AM.207} DEBUG ] Lock file /home/test/.ocpvscode.lock not found
[11:07:56 AM.676} INFO  ] Client connected
[11:07:56 AM.679} DEBUG ] Listener registered
```
**Screenshot (Explorer showing open folder "demo"):**
![vscode-ocp-cad-viewer-open demo folder](https://github.com/user-attachments/assets/2a02c94f-db7b-497b-84a8-57a3e78fca33)

**Screenshot (viewer and libraries managers)**
![vscode-ocp-cad-viewer-viewer and libraries managers](https://github.com/user-attachments/assets/1b885436-1bfb-4a0e-9ba8-af6b11359571)

**Comment by bernhard-42** [comment]
Can you check the developer tools of VS Code? It could be that you are running into a graphic driver issue here. We had that earlier with Linux

**Comment by WayneSherman** [comment]
It has something to do with opening a "folder" vs opening a file.  When I open a folder (which opens a new [workspace](https://code.visualstudio.com/docs/editor/workspaces#_how-do-i-open-a-vs-code-workspace)), it doesn't work.  But when I open the file directly, it works as expected.

After opening a file:
![vscode-ocp-cad-viewer-open demo file only (no workspace)](https://github.com/user-attachments/assets/ecb59d01-cdbe-4117-86b5-deae819aa2c5)

vscode open options:
![vscode-ocp-cad-viewer-open options](https://github.com/user-attachments/assets/ca85799c-7ab4-4e7e-88d1-ddd71959f575)

**Comment by WayneSherman** [comment]
I just discovered the problem can also happen when opening the file directly.  I think it is some kind of a race condition.  Opening a folder changes the timing and causes it to fail every time (so far).  Opening a file works more often, but can also fail.

My test system is on the slow side (running in VirtualBox) so it probably is exposing this race condition where a faster system would not see it.

**Comment by bernhard-42** [comment]
The viewer opens for example when you load or save a Python file that imports build123d or cadquery. If you close the viewer pane and then save the Python file again, does the viewer start again properly?

**Reply by WayneSherman** [reply]
>  If you close the viewer pane and then save the Python file again, does the viewer start again properly?

When I close the viewer pane and restart the viewer (either by saving the demo file or by clicking the "open viewer button"), sometimes the viewer opens properly, and other times it does not.  This is the case whether I have opened a "folder" or a single file.

The only difference between an open workspace (folder) and an open file is that the open workspace fails most of the time and the open file works most of the time.  But both cases have intermittent work/fail behavior.  I suspect there is some kind of an initialization race condition.

**Comment by bernhard-42** [comment]
hmm, interesting.
There should be some error message somewhere ...
In VS Code the "output" pane should have a "OCP CAD Viewer" section, is there anything visible?
And the next place are the developer tools (iirc, shift-ctrl-p and the "toggle developer tools"), it's basically the electron browser console.
It would be so much easier to understand if good case and bad case would somehow differentiate.
If there is nothing, I need my computer next week to maybe get an idea ...

**Comment by WayneSherman** [comment]
Developer Tools Console - CAD Viewer Failed
[vscode-app-1727287296578-file_only-cad_viewer_failed.log](https://github.com/user-attachments/files/17136132/vscode-app-1727287296578-file_only-cad_viewer_failed.log)

Developer Tools Console - CAD Viewer Worked
[vscode-app-1727287222606-file_only-cad_viewer_worked.log](https://github.com/user-attachments/files/17136126/vscode-app-1727287222606-file_only-cad_viewer_worked.log)

The logs are exactly the same up to line 33.
Lines 34 and 35 are basically the same (differ in IDs only)

Console Log Lines 36-39 in failed case:
```
xterm.js:1 [Violation] 'requestAnimationFrame' handler took 97ms
[Violation] Forced reflow while executing JavaScript took 53ms
VM10:601 resize listener registered
VM10:767 message listener registered
[End-of-file]
```

Console Log Lines 36-38 in working case:
```
VM10:601 resize listener registered
VM10:767 message listener registered
VM10:358 three-cad-viewer: WebGL Renderer created
... log continues with successful init
```

**Comment by bernhard-42** [comment]
Thanks @WayneSherman 
That gives me an idea where to look for. The viewer gets properly initialized, but the showViewer command is not triggered. Maybe the show event gets lost on your system. But now I need my computer to do some tests. Will look at it next week.
